### DISCUSSION TITLE
Inheritance mapping

### DISCUSSION THREAD

**Discussion by vieiralvitesse** [riok/mapperly#299] [Q&A]
Hi,

I read through the docs but couldn't find a way to map Inherited objects.

What would be the best way of achieving this?

thanks so much!

Leo

**Comment by latonz** [comment]
This should "just work". Could you provide an example, where it doesn't behave as expected?

**Reply by vieiralvitesse** [reply]
Sure, I might be doing something wrong, but what I want is:

Given an inheritance, when I have subtype entities I want the mapper to translate them to subtype DTO, the current behavior (from the snippet) is: the child properties are being discarded since I'm not specifying the type explicitly (I can't specify since my contracts expect a base parent class so the mapper has to check for the runtime type)

```using Riok.Mapperly.Abstractions;

var childDTO = PersonMapper.Map(new Child { Age = 10, Name = "John" });
Console.WriteLine(childDTO.GetType());

public class Person { public int Age { get; set; } }
public class Child : Person { public string Name { get; set; } }
public class PersonDTO { public int Age { get; set; } }
public class ChildDTO : PersonDTO { public string Name { get; set; } }

[Mapper]
public static partial class PersonMapper
{
    public static partial PersonDTO Map(Person person);
}

```

the output from console is `PersonDTO` instead of `ChildDTO`

**Reply by latonz** [reply]
This is not possible, but you can simply add a `public static partial ChildDTO Map(Child child);` method to your mapper. This method should then include mappings of properties from the person and the child.

**Reply by vieiralvitesse** [reply]
I see, unfortunately I can't do that since I don't know the type at compile type, from my experience with other mappers they also don't have this capability out of the box unless you explicitly specify the inheritance structure, otherwise the mapper is oblivious of the derived types.

thanks very much for your help and quick replies, I'll keep an eye on the following releases and Mapperly will definitely be my tool of choice moving forward!

**Reply by latonz** [reply]
No worries. If the subtypes are not known at compile time, there is no way Mapperly can generate code for the type at compile time. Since Mapperly is a source generator this won't work.

**Reply by vieiralvitesse** [reply]
btw, there's a workaround for this, we can create a user defined mapping and go through the inheritance manually:

```csharp
[Mapper]
public static partial class ContainerMapper
{
    public static partial ContainerDTO Map(Container container);

    private static InheritanceDTO MapInheritance(Inheritance inheritance)
    {
        return inheritance switch
        {
            FirstDerivative _ => new FirstDerivativeDTO(),
            SecondDerivative _ => new SecondDerivativeDTO(),
            _ => throw new ArgumentException()
        };
    }
}
```
